classes/controller/FrontController.php

```diff
@@ -465,7 +472,12 @@ public function init()
         $this->context->cart = $cart;
         $this->context->currency = $currency;
 
-        Hook::exec('actionFrontControllerAfterInit');
+        Hook::exec(
+            'actionFrontControllerInitAfter',
+            [
+                'controller' => $this,
```

Comment:
@PierreRambaud don't we have an access to `$this->context->controller` here? is there any point to provide `controller` parameter?

Reply:
Always relying on the context is not a good practice, this is our main struggle in fixing so many bugs because it forces us to override the context, or mock the context, or init a missing context, ...

Imho hooks, like Symfony events, should provide the parameters related to their scope so that they can be handled independently and not relying on the context (or the least possible)

Reply:
@jolelievre thanks, interesting point, I'd be happy to read more about possible issues when it comes to `$this->context->controller` vs `$params['controller']` when it comes to application which is already initialized and where `Context` should be our center of truth 👍

Reply:
We should only trust what we have in parameters. The Context can be easily modified by everything.

Reply:
Context has been the center of truth for years indeed, but the result is that everybody is used to access and, more dangerous, modify it (even inside the core code, it's not just in modules)

The problem is that all of the legacy code relies on the context in every single piece of code, meaning we can't reuse the legacy classes without having perfectly initialized or duplicated the Context Which makes it very hard to reuse, and even harder to test

So of course you will get the same thing in `$this->context->controller` and in `$params['controller']` but since the goal in the future is to remove the Context it's better to improve the rest of the code piece by piece so that we progressively stop relying on it

Reply:
@PierreRambaud yeah but like... entire PrestaShop and its modules are relying on `Context`, what else do you recommend for checking current `Shop`, `Currency`, `Language`? :) We have `Context` injected everywhere in Symfony services 👀 

We literally don't have any hooks where you give parameters explicitly, even entire `Cart` mechanism is based on `Context`, not only in legacy part of code.

I'm just asking for anyone visiting this PR and for myself how can we approach this issue, what's the better way of getting all informations about current `state` of PrestaShop app, currency etc. We can't have immutable state, we need to be able to change `Context` in some scenarios :) 

Also I don't really see a way to override `$this->context->controller` in this place but that doesn't matter 👍 

I'd be happy to discuss this on Slack or something because you and @jolelievre give a pretty controversial advice ;)

Reply:
Yes, it should be discussed on slack or in another issue. :sweat_smile:

Reply:
You're right we rely on it everywhere and it's become the perfect shortcut, I'm not saying we can stop using it today but most of the object you mentioned have alternatives (some are already possible, some not, it's just to explain the ideal way):

- Currency should be fetched based on the Cart or Order you are editing
  - if it's to display something on front it should be fetched from the Customer's preferences
- Language should be fetched based on the employee or customer in the session
- Shop is trickier ^^ Sometimes it indeed depends on the context, sometimes (like when you edit an order) you shouldn't use the Context shop at all, but the one that was associated to the Order on creation

The thing is most of the time we don't even think of what we really need (should I use the shop assigned to the cart/order or the one from the context) and the context value is always used by default But it's not always the appropriate one Which is why the singleton approach has many flows

So it might seem controversial, sometimes you have no choice but to use the Context, but when given the possibility to do otherwise (especially when parameters adapted to the current action are provided) it should be favored versus using the Context

Now we clearly lack some proper alternatives Especially in Symfony services initialization where we still heavily rely on LegacyContext to inject language, shop or other stuffs However we always **try** not to inject Context directly, but rather the info we need specifically (like default language, or current shop context)

The road is long but we can still try to improve it ^^